In [ ]:
#必须使用A100才可以
!git clone https://github.com/THUDM/GLM-4.git

Cloning into 'GLM-4'...
remote: Enumerating objects: 1480, done.
remote: Counting objects: 100% (362/362), done.
remote: Compressing objects: 100% (162/162), done.
remote: Total 1480 (delta 254), reused 201 (delta 200), pack-reused 1118 (from 2)
Receiving objects: 100% (1480/1480), 47.73 MiB | 16.79 MiB/s, done.
Resolving deltas: 100% (807/807), done.


In [ ]:
!pip install -r /content/GLM-4/basic_demo/requirements.txt

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.1/18.1 MB 86.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.7/318.7 kB 31.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 105.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 74.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 53.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 10.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 40.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 18.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 3.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 95.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
!pwd

/content


In [ ]:
!ls

GLM-4  sample_data  test.png


In [ ]:
import torch
from PIL import Image
from transformers import AutoModelForCausalLM, AutoTokenizer

device = "cuda"

tokenizer = AutoTokenizer.from_pretrained("THUDM/glm-4v-9b", trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
    "THUDM/glm-4v-9b",
    torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=True,
    trust_remote_code=True
).to(device).eval()



gen_kwargs = {"max_length": 2500, "do_sample": True, "top_k": 1}




/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/3.22k [00:00<?, ?B/s]

tokenization_chatglm.py:   0%|          | 0.00/17.7k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- tokenization_chatglm.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


tokenizer.model:   0%|          | 0.00/2.62M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.77k [00:00<?, ?B/s]

configuration_chatglm.py:   0%|          | 0.00/2.57k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- configuration_chatglm.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_chatglm.py:   0%|          | 0.00/57.0k [00:00<?, ?B/s]

visual.py:   0%|          | 0.00/7.00k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- visual.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.
A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- modeling_chatglm.py
- visual.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors.index.json:   0%|          | 0.00/111k [00:00<?, ?B/s]

model-00001-of-00015.safetensors:   0%|          | 0.00/1.95G [00:00<?, ?B/s]

model-00002-of-00015.safetensors:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

model-00003-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00004-of-00015.safetensors:   0%|          | 0.00/1.93G [00:00<?, ?B/s]

model-00005-of-00015.safetensors:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

model-00006-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00007-of-00015.safetensors:   0%|          | 0.00/1.93G [00:00<?, ?B/s]

model-00008-of-00015.safetensors:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

model-00009-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00010-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00011-of-00015.safetensors:   0%|          | 0.00/1.96G [00:00<?, ?B/s]

model-00012-of-00015.safetensors:   0%|          | 0.00/1.98G [00:00<?, ?B/s]

model-00013-of-00015.safetensors:   0%|          | 0.00/1.96G [00:00<?, ?B/s]

model-00014-of-00015.safetensors:   0%|          | 0.00/1.98G [00:00<?, ?B/s]

model-00015-of-00015.safetensors:   0%|          | 0.00/811M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/15 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/205 [00:00<?, ?B/s]

In [ ]:
query = '描述这张图片'
image = Image.open("test.png").convert('RGB')
inputs = tokenizer.apply_chat_template([{"role": "user", "image": image, "content": query}],
                                       add_generation_prompt=True, tokenize=True, return_tensors="pt",
                                       return_dict=True)  # chat mode

inputs = inputs.to(device)


with torch.no_grad():
    outputs = model.generate(**inputs, **gen_kwargs)
    outputs = outputs[:, inputs['input_ids'].shape[1]:]
    print(tokenizer.decode(outputs[0]))

图片展示了三个Windows系统的更新补丁，这些更新补丁的版本号分别为2024、5.130和2024-11，它们都适用于Windows 11系统，并针对不同的安全漏洞和稳定性问题进行了修复。第一个更新补丁KB5046633正在等待重启，另外两个更新补丁KB890830和KB5045935都可以安装。 <|endoftext|>


In [ ]:

def count_parameters(model):
    return sum(p.numel() for p in model.parameters())

# 假设你已经定义或加载了 ChatGLMForConditionalGeneration 模型
# model = ChatGLMForConditionalGeneration(...)

# 计算总参数量
total_params = count_parameters(model)
print(f"模型的总参数量为: {total_params:,}")


模型的总参数量为: 13,906,320,896


In [ ]:
query = '描述这张图片'
image = Image.open("/content/test1.jpg").convert('RGB')
inputs = tokenizer.apply_chat_template([{"role": "user", "image": image, "content": query}],
                                       add_generation_prompt=True, tokenize=True, return_tensors="pt",
                                       return_dict=True)  # chat mode

inputs = inputs.to(device)


with torch.no_grad():
    outputs = model.generate(**inputs, **gen_kwargs)
    outputs = outputs[:, inputs['input_ids'].shape[1]:]
    print(tokenizer.decode(outputs[0]))

这是一张女性侧身照，她穿着白色吊带，披着头发，表情平静地看着镜头。背景是海边的礁石和波涛汹涌的大海，远处可以看到山脉的轮廓。在图片的右下角有一个水印，写着“知乎@天狼”。 <|endoftext|>


In [ ]:
#图片输入尺寸为1120*1120，经过14*14卷积核，变为6400块
model

ChatGLMForConditionalGeneration(
  (transformer): ChatGLMModel(
    (embedding): Embedding(
      (word_embeddings): Embedding(151552, 4096)
    )
    (rotary_pos_emb): RotaryEmbedding()
    (encoder): GLMTransformer(
      (layers): ModuleList(
        (0-39): 40 x GLMBlock(
          (input_layernorm): RMSNorm()
          (self_attention): SelfAttention(
            (query_key_value): Linear(in_features=4096, out_features=4608, bias=True)
            (core_attention): CoreAttention(
              (attention_dropout): Dropout(p=0.0, inplace=False)
            )
            (dense): Linear(in_features=4096, out_features=4096, bias=False)
          )
          (post_attention_layernorm): RMSNorm()
          (mlp): MLP(
            (dense_h_to_4h): Linear(in_features=4096, out_features=27392, bias=False)
            (dense_4h_to_h): Linear(in_features=13696, out_features=4096, bias=False)
          )
        )
      )
      (final_layernorm): RMSNorm()
    )
    (output_layer): Linear(in

`ChatGLMForConditionalGeneration` 模型的结构复杂且层次丰富，包含文本生成（语言模型）和视觉处理（EVA2CLIP）两个主要部分，分别用于文本生成和视觉信息处理。这种模型是典型的多模态大模型架构，既能处理文本，又能理解视觉输入。下面我将逐层解析其结构。

### 1. **整体架构**
   - 模型主要由一个文本生成模块 (`ChatGLMModel`) 和一个视觉处理模块 (`EVA2CLIPModel`) 组成。
   - `ChatGLMModel` 负责处理文本输入，生成条件输出。
   - `EVA2CLIPModel` 负责视觉输入的处理，使模型能够结合图像信息进行推理和生成。

### 2. **ChatGLMModel 模块**
   - **(embedding): Embedding**：
     - 包含 `word_embeddings` 层，维度为 `(151552, 4096)`，即模型的词汇表大小为 151552，每个词向量的维度为 4096。
     - `word_embeddings` 主要作用是将离散的词表示为连续的向量，便于后续的神经网络处理。

   - **(rotary_pos_emb): RotaryEmbedding**：
     - 使用了旋转位置编码（Rotary Embedding），这是一种用于自注意力机制的相对位置编码方式，有助于捕获词与词之间的顺序信息。

   - **(encoder): GLMTransformer**：
     - 这是一个基于 Transformer 的编码器，用于将输入文本编码成上下文表示。
     - `GLMTransformer` 包含 40 层 `GLMBlock`，每个 `GLMBlock` 包含自注意力和前馈神经网络层。

       - **GLMBlock**:
         - 每个 `GLMBlock` 包括以下组件：
           - **(input_layernorm): RMSNorm**：采用 RMSNorm 归一化层，避免传统 LayerNorm 中均值偏移的计算，RMSNorm 常用于提高模型的稳定性。
           - **(self_attention): SelfAttention**：自注意力机制。
             - `(query_key_value): Linear`：用于生成 `query`、`key` 和 `value` 矩阵，输入维度为 4096，输出维度为 4608，包含偏置。
             - **(core_attention): CoreAttention**：核心的注意力机制，包含一个 dropout 层。
             - **(dense): Linear**：线性层，注意力输出的变换层，将输出维度重投影为 4096。
           - **(post_attention_layernorm): RMSNorm**：注意力后的归一化层。
           - **(mlp): MLP**：前馈神经网络层，常见于 Transformer 结构中。
             - `(dense_h_to_4h)`：线性层，将 4096 维变换为 27392 维。
             - `(dense_4h_to_h)`：线性层，将 13696 维重新变换为 4096 维。
     - **(final_layernorm): RMSNorm**：最后一层归一化层。

   - **(output_layer): Linear**：
     - 输出层，将 4096 维的隐藏状态映射到词汇表大小（151552）的输出维度，用于生成最终的文本。

### 3. **EVA2CLIPModel 模块（视觉处理部分）**
   - **(patch_embedding): PatchEmbedding**：
     - 图像的补丁嵌入，用于将输入的图像切分成补丁（patch），类似于 ViT 模型。
     - **(proj): Conv2d**：二维卷积层，将 RGB 图像切分成尺寸为 1792 的特征补丁，卷积核和步幅均为 `(14, 14)`。
     - **(position_embedding): Embedding**：位置嵌入层，输入大小为 6401，每个位置嵌入向量的维度为 1792。

   - **(transformer): Transformer**：
     - 视觉 Transformer 模块，共有 63 层 `TransformerLayer`，用于处理图像补丁嵌入。
     - **TransformerLayer**：
       - **(input_layernorm): LayerNorm**：归一化层，作用于 1792 维的特征。
       - **(attention): Attention**：注意力层，结构与文本生成部分类似。
         - `(query_key_value): Linear`：线性层，输入维度为 1792，输出维度为 5376。
         - `(dense): Linear`：线性层，维度为 1792，用于输出投影。
       - **(mlp): MLP**：前馈神经网络。
         - 激活函数为 `GELU`。
         - 包含两层线性变换，将 1792 维映射到 15360 维再映射回 1792 维。
       - **(post_attention_layernorm): LayerNorm**：注意力后的归一化层。

   - **(linear_proj): GLU**：
     - GLU（Gated Linear Unit）层，用于线性投影，将图像特征转化为 4096 维。
     - 包括一个 `LayerNorm`、`GELU` 激活函数以及多个线性层（`dense_h_to_4h` 和 `dense_4h_to_h`）。

   - **(conv): Conv2d**：
     - 卷积层，将图像特征的通道数从 1792 变换到 4096，卷积核大小为 `(2, 2)`，步幅也为 `(2, 2)`。

### 4. **总结**
   - `ChatGLMForConditionalGeneration` 是一个复杂的多模态模型，包含文本处理（GLM Transformer）和视觉处理（EVA2CLIP Transformer）两大模块。
   - 文本处理模块使用自注意力、前馈神经网络、归一化等层堆叠而成，类似于传统 Transformer，但引入了 RMSNorm 和 Rotary Embedding 等改进。
   - 视觉处理模块通过将图像切分为补丁，再经过多层 Transformer 层来提取图像特征，最后通过 GLU 和卷积层对图像特征进行进一步转换。
   - 这种架构使得模型能够处理和理解文本和图像两种模态，适用于多模态生成和理解任务。

这类多模态模型在 AI 应用中具有广泛的潜力，尤其适用于文本生成、图像理解和多模态交互等复杂任务。

在这个模型中，当你输入一张图片并希望它输出文字描述时，模型会使用其视觉处理模块 (`EVA2CLIPModel`) 将图像信息转换为可用于文本生成的特征向量，然后通过文本生成模块 (`ChatGLMModel`) 将这些特征转化为自然语言的文字描述。下面是详细的处理过程：

### 1. **图像预处理与特征提取**
   - 首先，模型将输入的图像进行预处理，将其分割成一系列图像补丁（patch）。
   - **Patch Embedding**：
     - 图像进入 `EVA2CLIPModel` 中的 `PatchEmbedding` 层。这里通过卷积操作（`Conv2d`）将图像划分成多个 14x14 的图像补丁，并将这些补丁映射到一个高维空间（1792 维）。
     - 每个图像补丁都会加上一个位置嵌入 (`position_embedding`)，用于保留图像中空间位置的信息，以便模型能理解图像的结构和布局。
   
   - **Transformer Layers**：
     - 接着，嵌入后的图像补丁序列会经过 `EVA2CLIPModel` 中的 63 层 `TransformerLayer`，这些层通过自注意力机制和前馈神经网络对图像特征进行多层次的抽象和提取。
     - 自注意力机制帮助模型捕捉图像中的全局关系，例如物体之间的空间关系、背景和前景的区分等。

   - **Feature Projection**：
     - 经过多层 Transformer 的处理后，图像的特征会被进一步处理，通过 `linear_proj`（GLU层）和 `conv`（卷积层）将这些特征转换为一个 4096 维的向量。这一向量包含了图像的全局语义信息，并被作为图像的特征表示。

### 2. **视觉特征到语言特征的转换**
   - **连接视觉和语言模块**：
     - 提取到的图像特征向量（4096 维）会作为 `ChatGLMModel` 中 Transformer 的上下文输入。这样，视觉信息就能参与到文本生成过程中。
     - 通常，图像特征会与语言模型的初始隐状态（或特定标记）结合起来，以便将视觉信息注入到生成的文本流中。

### 3. **文本生成模块 (ChatGLMModel)**
   - **输入嵌入**：
     - `ChatGLMModel` 主要用于生成文本描述。在这个阶段，模型会逐步生成描述文字，并在生成过程中参考图像特征。
     - 每一步生成的词会嵌入到高维空间，并通过自注意力和上下文信息来生成下一个词，逐步构建完整的描述。

   - **多轮自注意力机制**：
     - 在 `ChatGLMModel` 中的 40 层 Transformer 块中，自注意力机制可以帮助模型在生成文本时不断参考图像信息。这些注意力机制会引导模型在生成特定单词时聚焦于相关的视觉信息。
     - 例如，当模型生成“天空”这个词时，自注意力机制可以将焦点引导到图像中的天空区域，从而生成更符合图像内容的描述。

   - **语言解码**：
     - 每一层的处理完成后，最后一层的输出经过 `output_layer` 映射回词汇表的维度，通过 Softmax 层选择下一个生成的词。
     - 通过这种迭代生成的方式，模型逐步生成文字描述，直到达到结束条件（例如生成了句子结束标记）。

### 4. **最终输出**
   - 最终生成的词序列就是对输入图像的自然语言描述。这一描述既包含了图像中的主要物体信息，也可能包括模型推测出的场景、情感或意图等高级语义信息。

### 总结
   - **视觉部分** (`EVA2CLIPModel`) 提取图像特征并将其编码为一个语义向量。
   - **语言部分** (`ChatGLMModel`) 在生成文本的过程中利用视觉特征，生成符合图像内容的文字描述。
   - 模型的这种结构使得它能够从视觉到语言的多模态转换，因此它可以在图像输入的基础上生成符合图像内容的文字描述。

这种多模态模型的优势在于它能将图像和文本特征结合，适用于视觉描述生成、图文检索和多模态问答等任务。

在多模态大模型中，当视觉特征与文本特征融合后进入注意力机制进行计算时，**Query (Q)**、**Key (K)**、**Value (V)** 的选择通常取决于当前计算的任务和模型的结构。在多模态融合中，常见的注意力设计方式主要包括**跨模态注意力**和**自注意力**，而它们对 \( Q \)、\( K \)、\( V \) 的选择有所不同。

以下是常见的多模态注意力设计和其 Q、K、V 的含义和选择：

### 1. **跨模态注意力（Cross-Modal Attention）**
   跨模态注意力用于在一个模态的特征上“查询”另一个模态的特征，典型应用场景是文本对图像的注意力（如描述生成），或图像对文本的注意力（如视觉问答）。

   - **Q**：通常来自主模态（例如，文本特征或视觉特征），该模态在当前任务中需要重点处理。
   - **K** 和 **V**：来自辅助模态的特征（例如，当文本描述生成时，K 和 V 来自视觉特征；当图像生成或分类时，K 和 V 来自文本特征）。
   
   这种方式允许模型通过关注辅助模态来丰富主模态的信息。例如：
   - 在描述生成任务中，**Q** 是文本嵌入，**K** 和 **V** 是视觉嵌入。文本可以根据图像内容动态生成描述。
   - 在视觉问答中，**Q** 是问题文本嵌入，**K** 和 **V** 是图像嵌入，模型通过关注图像特定区域来生成回答。

### 2. **自注意力（Self-Attention）**
   当模型在融合后的多模态特征上执行自注意力时，**Q**、**K** 和 **V** 通常来自同一个融合特征序列。这种融合方式尤其适用于视觉和文本的深层次交互，例如通过多层 Transformer 实现跨模态信息共享。

   - **Q**、**K**、**V** 均为融合后的特征。
   - 这种设计允许文本和图像的特征相互联系，模型在每一层中通过自注意力机制选择性关注两个模态的相关信息。
   
   在这种情境下，视觉和文本特征在注意力机制中彼此相互关联、交互融合，形成更具上下文信息的表示。例如，文本嵌入可“参考”图像特征中的相关内容来调整其自身表示。

### 3. **双向注意力（Bidirectional Attention）**
   一些多模态模型中，也采用**双向注意力**（Bidirectional Attention），即分别在文本对图像和图像对文本的特征上各自计算一次跨模态注意力，最终融合两个方向的输出。

   - 第一层计算时，**Q** 来自文本嵌入，**K** 和 **V** 来自视觉嵌入（文本对图像的注意力）。
   - 第二层计算时，**Q** 来自视觉嵌入，**K** 和 **V** 来自文本嵌入（图像对文本的注意力）。
   
   双向注意力用于捕捉两个模态之间的双向信息交互，最终输出的特征会结合来自文本和图像的双向关系。

### 总结
- **Q**：通常来自需要“关注”另一个模态的特征，代表当前模态的查询向量。
- **K 和 V**：一般来自被关注的模态，提供上下文或信息供 Q 查询使用。
  
这种设计允许多模态模型根据任务动态选择性地关注不同模态的信息，提高模型在多模态任务中的表现。

在多模态大模型中，视觉特征和文本特征的融合是核心步骤，旨在使模型能够有效地理解和处理来自不同模态的信息。这些融合方法可以大致分为以下几种：

### 1. **简单拼接（Concatenation）**
   将视觉特征和文本特征直接拼接起来，然后通过一个或多个全连接层进行变换。这种方法直观而简单，但可能无法充分捕捉到两个模态之间的复杂关系。

   - **适用场景**：简单的多模态分类任务或需要快速实验的场景。
   - **优缺点**：简单易实现，但无法捕捉复杂的模态间依赖关系。

### 2. **线性变换拼接（Linear Projection Fusion）**
   先通过线性层或 MLP 将视觉和文本特征分别映射到相同的维度空间，再将它们拼接在一起。这可以使得两个模态在一个统一的表示空间中融合。

   - **适用场景**：较为简单的多模态任务，如图像标题生成或分类。
   - **优缺点**：较拼接方法稍微增强了模态间的互动，但对模态之间的复杂交互仍然缺乏较强的建模能力。

### 3. **跨模态注意力（Cross-Modal Attention）**
   通过跨模态注意力机制，一个模态作为**查询（Query）**，另一个模态作为**键（Key）**和**值（Value）**，计算注意力权重，以实现特征之间的交互和关联。这种方法可以有效地让每个模态关注另一个模态的细节信息。

   - **适用场景**：图像描述生成、视觉问答等需要显著模态间互动的任务。
   - **优缺点**：能精确捕捉模态间的复杂关系，但计算开销较大。

### 4. **自注意力（Self-Attention on Combined Features）**
   将视觉特征和文本特征按顺序组合，然后在组合后的特征上执行自注意力机制。视觉和文本特征会在同一层级上相互作用，能够捕捉较深层次的模态间关系。

   - **适用场景**：需要深层模态融合的任务，如图像生成、语义分割等。
   - **优缺点**：能深入捕捉模态间的相互关系，但要求模型有较强的计算资源。

### 5. **循环交互（Iterative Interaction）**
   多轮次交替地进行视觉和文本特征之间的交互。每一轮中，视觉特征可以更新文本特征，文本特征反过来更新视觉特征。这种交互方式能强化两个模态的相互理解，逐步增强多模态信息的表达。

   - **适用场景**：需要逐步生成或完善的多模态任务，如多轮对话生成。
   - **优缺点**：迭代增强了模态间的关系，但计算时间较长。

### 6. **双流模型（Dual-Stream Model）**
   使用两个独立的 Transformer 分别对视觉和文本特征进行编码，然后通过特定层（如交叉注意力层或共享的最终层）来实现两个流的交互。典型的实现如 CLIP 模型，将视觉和文本特征编码为同一特征空间中的向量。

   - **适用场景**：需要多模态嵌入对齐的任务，如图文检索和匹配。
   - **优缺点**：可以在多个模态之间生成一致的嵌入表示，但实现和训练较为复杂。

### 7. **门控机制（Gating Mechanism）**
   使用门控机制（如门控单元或门控注意力）对视觉和文本特征进行选择性融合。门控单元可以根据输入特征的权重，选择性地增强或抑制某一模态的特征，以避免信息冗余或噪声的影响。

   - **适用场景**：需要选择性融合信息的任务，如文本引导的目标检测或多模态情感分析。
   - **优缺点**：可以有效处理冗余信息，但设计和调优门控机制较复杂。

### 8. **共注意力（Co-Attention）**
   使用双向的跨模态注意力机制，即图像对文本进行注意力计算，文本对图像进行注意力计算，获得互相影响的特征。这种机制允许模型通过两个模态的相互注意来精确定位两个模态之间的关联。

   - **适用场景**：图文问答、图文匹配任务。
   - **优缺点**：可以实现高精度的模态间关系建模，但通常对内存和计算要求较高。

### 9. **对比学习（Contrastive Learning）**
   将视觉和文本特征分别映射到相同的特征空间，通过对比学习将相同语义的图像-文本对聚集到一起，而不同语义的图像-文本对则被分散。典型的实现是 CLIP 模型，通过对比损失来增强图文匹配的能力。

   - **适用场景**：图文检索、图文匹配任务。
   - **优缺点**：适合生成模态不对齐的通用特征空间，但仅通过对比学习可能不能完全捕获细粒度的模态交互。

### 总结
在实际应用中，这些方法通常会结合使用，例如跨模态注意力和自注意力交替使用，以增强两个模态的交互表达能力。选择合适的融合方法要根据任务的需求、数据特点以及计算资源的约束等因素来决定。